# Gathering Data

For this project, I will gather data from three different sources.

In [1]:
# This extension formats all cells in the notebook using black
%load_ext nb_black

<IPython.core.display.Javascript object>

## The first source
The `twitter-archive-enhanced.csv` which was provided by *Udacity*.

This file contains the twitter archive of the [WeRateDogs twitter account](https://twitter.com/dog_rates?ref_src=twsrc%5Egoogle%7Ctwcamp%5Eserp%7Ctwgr%5Eauthor) which is the "only source for professional dog ratings" as they describe themselves on their twitter account.

I downloaded this file manually and added it to the folder of the project under its original name.

![](./Slides/Manual_Download_Proof.png)

##### Storing downloaded data into a pandas DataFrame

In [2]:
import pandas as pd

tweet_archive_df = pd.read_csv("twitter-archive-enhanced.csv")

<IPython.core.display.Javascript object>

## The second source

The `image_predictions.tsv` which is hosted on *Udacity* servers. This file contains the tweet image predictions, i.e., what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network.

##### Downloading the file programmatically using *requests* Python module.

In [3]:
import requests

url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
predictions_file_name = url.rsplit("/")[-1]

response = requests.get(url)

with open(predictions_file_name, "wb") as saved_file:
    saved_file.write(response.content)

<IPython.core.display.Javascript object>

![](./Slides/Programmatic_Download_Proof.png)

##### Storing the collected data in a pandas DataFrame

In [4]:
img_pred_df = pd.read_csv(predictions_file_name, sep="\t")

<IPython.core.display.Javascript object>

## The third source

Using the `tweet_id` column in the WeRateDogs Twitter archive `twitter-archive-enhanced.csv`, I will query the Twitter API for each tweet's JSON data using Python's *Tweepy* module and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data will be written to its own line.

##### Loading secret API keys from environment

In [5]:
import os
from dotenv import load_dotenv

load_dotenv()

CONSUMER_KEY = os.getenv("CONSUMER_KEY")
CONSUMER_SECRET = os.getenv("CONSUMER_SECRET")
ACCESS_TOKEN = os.getenv("ACCESS_TOKEN")
ACCESS_SECRET = os.getenv("ACCESS_SECRET")

<IPython.core.display.Javascript object>

##### Querying the Twitter API through Tweepy to collect tweet data

In [6]:
# import tweepy
# import json

# # Authentication
# auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
# auth.set_access_token(ACCESS_TOKEN, ACCESS_SECRET)

# api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

# tweet_ids = pd.read_csv("twitter-archive-enhanced.csv").tweet_id

# with open("tweet-json.txt", "w") as txt_file:
#     # For every tweet ID
#     for tweet_id in tweet_ids:
#         # Try to get the tweet information
#         try:
#             tweet = api.get_status(tweet_id, tweet_mode="extended")
#         except tweepy.TweepError:
#             continue

#         # If successful, write the json string to the tweet-json.txt file and append a new line
#         tweet_json_string = json.dumps(tweet)
#         txt_file.write(f"{tweet_json_string}\n")

<IPython.core.display.Javascript object>

![](./Slides/API_Querying_proof.png)

##### Saving collected tweets data into a pandas DataFrame

In [7]:
tweet_data_df = pd.DataFrame()
with open("tweet-json.txt", "r") as txt_file:
    tweet_data_df = pd.read_json(txt_file, lines=True)

<IPython.core.display.Javascript object>

In [8]:
tweet_data_df

created_at                  id              id_str  \
0    2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   
1    2017-08-01 00:17:27+00:00  892177421306343426  892177421306343424   
2    2017-07-31 00:18:03+00:00  891815181378084864  891815181378084864   
3    2017-07-30 15:58:51+00:00  891689557279858688  891689557279858688   
4    2017-07-29 16:00:24+00:00  891327558926688256  891327558926688256   
...                        ...                 ...                 ...   
2349 2015-11-16 00:24:50+00:00  666049248165822465  666049248165822464   
2350 2015-11-16 00:04:52+00:00  666044226329800704  666044226329800704   
2351 2015-11-15 23:21:54+00:00  666033412701032449  666033412701032448   
2352 2015-11-15 23:05:30+00:00  666029285002620928  666029285002620928   
2353 2015-11-15 22:32:08+00:00  666020888022790149  666020888022790144   

                                              full_text  truncated  \
0     This is Phineas. He's a mystical boy. Only eve...      False   
1     This is Tilly. She's just checking pup on you....      False   
2     This is Archie. He is a rare Norwegian Pouncin...      False   
3     This is Darla. She commenced a snooze mid meal...      False   
4     This is Franklin. He would like you to stop ca...      False   
...                                                 ...        ...   
2349  Here we have a 1949 1st generation vulpix. Enj...      False   
2350  This is a purebred Piers Morgan. Loves to Netf...      False   
2351  Here is a very happy pup. Big fan of well-main...      False   
2352  This is a western brown Mitsubishi terrier. Up...      False   
2353  Here we have a Japanese Irish Setter. Lost eye...      False   

     display_text_range                                           entities  \
0               [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1              [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2              [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3               [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                 ...                                                ...   
2349           [0, 120]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2350           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2351           [0, 130]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2352           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2353           [0, 131]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
0     {'media': [{'id': 892420639486877696, 'id_str'...   
1     {'media': [{'id': 892177413194625024, 'id_str'...   
2     {'media': [{'id': 891815175371796480, 'id_str'...   
3     {'media': [{'id': 891689552724799489, 'id_str'...   
4     {'media': [{'id': 891327551943041024, 'id_str'...   
...                                                 ...   
2349  {'media': [{'id': 666049244999131136, 'id_str'...   
2350  {'media': [{'id': 666044217047650304, 'id_str'...   
2351  {'media': [{'id': 666033409081393153, 'id_str'...   
2352  {'media': [{'id': 666029276303482880, 'id_str'...   
2353  {'media': [{'id': 666020881337073664, 'id_str'...   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2349  <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r..

<IPython.core.display.Javascript object>

## Wrap-up

In this section, I gathered data from 3 different sources.

- `twitter-archive-enhanced.csv` which is now loaded into **`tweet_archive_df`**
- Downloaded `image_predictions.tsv` programmatically from Udacity's servers and loaded into **`img_pred_df`**
- Queried the Twitter API through Tweepy Python module and loaded the data into **`tweet_data_df`**

A random sample of the rows from the three tables is shown below

In [9]:
tweet_archive_df.sample(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1088  737826014890496000                    NaN                  NaN   
2266  667530908589760512                    NaN                  NaN   

                      timestamp  \
1088  2016-06-01 02:00:04 +0000   
2266  2015-11-20 02:32:25 +0000   

                                                 source  \
1088  <a href="http://twitter.com/download/iphone" r...   
2266  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
1088  This is Remy. He has some long ass ears (proba...                  NaN   
2266  Meet Naphaniel. He doesn't necessarily enjoy h...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1088                       NaN                        NaN   
2266                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1088  https://twitter.com/dog_rates/status/737826014...                10   
2266  https://twitter.com/dog_rates/status/667530908...                10   

      rating_denominator       name doggo floofer pupper puppo  
1088                  10       Remy  None    None   None  None  
2266                  10  Naphaniel  None    None   None  None

<IPython.core.display.Javascript object>

In [10]:
img_pred_df.sample(2)

tweet_id                                          jpg_url  \
109  667885044254572545  https://pbs.twimg.com/media/CUTN5V4XAAAIa4R.jpg   
892  699413908797464576  https://pbs.twimg.com/media/CbTRPXdW8AQMZf7.jpg   

     img_num        p1   p1_conf  p1_dog                p2   p2_conf  p2_dog  \
109        1  malamute  0.088530    True  golden_retriever  0.087499    True   
892        1   Samoyed  0.517479    True          malamute  0.155935    True   

             p3   p3_conf  p3_dog  
109      muzzle  0.075008   False  
892  Eskimo_dog  0.095001    True

<IPython.core.display.Javascript object>

In [11]:
tweet_data_df

created_at                  id              id_str  \
0    2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   
1    2017-08-01 00:17:27+00:00  892177421306343426  892177421306343424   
2    2017-07-31 00:18:03+00:00  891815181378084864  891815181378084864   
3    2017-07-30 15:58:51+00:00  891689557279858688  891689557279858688   
4    2017-07-29 16:00:24+00:00  891327558926688256  891327558926688256   
...                        ...                 ...                 ...   
2349 2015-11-16 00:24:50+00:00  666049248165822465  666049248165822464   
2350 2015-11-16 00:04:52+00:00  666044226329800704  666044226329800704   
2351 2015-11-15 23:21:54+00:00  666033412701032449  666033412701032448   
2352 2015-11-15 23:05:30+00:00  666029285002620928  666029285002620928   
2353 2015-11-15 22:32:08+00:00  666020888022790149  666020888022790144   

                                              full_text  truncated  \
0     This is Phineas. He's a mystical boy. Only eve...      False   
1     This is Tilly. She's just checking pup on you....      False   
2     This is Archie. He is a rare Norwegian Pouncin...      False   
3     This is Darla. She commenced a snooze mid meal...      False   
4     This is Franklin. He would like you to stop ca...      False   
...                                                 ...        ...   
2349  Here we have a 1949 1st generation vulpix. Enj...      False   
2350  This is a purebred Piers Morgan. Loves to Netf...      False   
2351  Here is a very happy pup. Big fan of well-main...      False   
2352  This is a western brown Mitsubishi terrier. Up...      False   
2353  Here we have a Japanese Irish Setter. Lost eye...      False   

     display_text_range                                           entities  \
0               [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1              [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2              [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3               [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                 ...                                                ...   
2349           [0, 120]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2350           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2351           [0, 130]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2352           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2353           [0, 131]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
0     {'media': [{'id': 892420639486877696, 'id_str'...   
1     {'media': [{'id': 892177413194625024, 'id_str'...   
2     {'media': [{'id': 891815175371796480, 'id_str'...   
3     {'media': [{'id': 891689552724799489, 'id_str'...   
4     {'media': [{'id': 891327551943041024, 'id_str'...   
...                                                 ...   
2349  {'media': [{'id': 666049244999131136, 'id_str'...   
2350  {'media': [{'id': 666044217047650304, 'id_str'...   
2351  {'media': [{'id': 666033409081393153, 'id_str'...   
2352  {'media': [{'id': 666029276303482880, 'id_str'...   
2353  {'media': [{'id': 666020881337073664, 'id_str'...   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2349  <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r..

<IPython.core.display.Javascript object>

# Assessing Data

## tweets' archive table

### Visual assessment

In [12]:
tweet_archive_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

<IPython.core.display.Javascript object>

In [13]:
tweet_archive_df.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

<IPython.core.display.Javascript object>

In [14]:
tweet_archive_df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1704  680889648562991104                    NaN                  NaN   
1923  674082852460433408                    NaN                  NaN   
58    880935762899988482                    NaN                  NaN   
86    875747767867523072                    NaN                  NaN   
1665  682750546109968385                    NaN                  NaN   

                      timestamp  \
1704  2015-12-26 23:15:17 +0000   
1923  2015-12-08 04:27:30 +0000   
58    2017-06-30 23:47:07 +0000   
86    2017-06-16 16:11:53 +0000   
1665  2016-01-01 02:29:49 +0000   

                                                 source  \
1704  <a href="http://twitter.com/download/iphone" r...   
1923  <a href="http://twitter.com/download/iphone" r...   
58    <a href="http://twitter.com/download/iphone" r...   
86    <a href="http://twitter.com/download/iphone" r...   
1665  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1704  This is Cheesy. It's her birthday. She's patie...                  NaN   
1923  This is a Sagitariot Baklava mix. Loves her ne...                  NaN   
58    This is Louis. He's crossing. It's a big deal....                  NaN   
86    This is Goose. He's a womanizer. Cheeky as h*c...                  NaN   
1665  Meet Taco. He's a speckled Garnier Fructis. Lo...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1704                       NaN                        NaN   
1923                       NaN                        NaN   
58                         NaN                        NaN   
86                         NaN                        NaN   
1665                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1704  https://twitter.com/dog_rates/status/680889648...                 9   
1923  https://twitter.com/dog_rates/status/674082852...                11   
58    https://twitter.com/dog_rates/status/880935762...                13   
86    https://twitter.com/dog_rates/status/875747767...                13   
1665  https://twitter.com/dog_rates/status/682750546...                 9   

      rating_denominator    name doggo floofer pupper puppo  
1704                  10  Cheesy  None    None   None  None  
1923                  10       a  None    None   None  None  
58                    10   Louis  None    None   None  None  
86                    10   Goose  None    None   None  None  
1665                  10    Taco  None    None   None  None

<IPython.core.display.Javascript object>

### Programmatic assessment

In [15]:
tweet_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

<IPython.core.display.Javascript object>

In [16]:
tweet_archive_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

<IPython.core.display.Javascript object>

In [17]:
tweet_archive_df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

<IPython.core.display.Javascript object>

### Findings

**Quality Issues**

- Missing values in:
    - `in_reply_to_status_id`
    - `in_reply_to_user_id`
    - `retweeted_status_id`
    - `retweeted_status_user_id`
    - `retweeted_status_timestamp`
    - `expanded_urls`


- Outlier values in `rating_denominator`.

*It is understandable that rating numerators are through the roof because let's face it, dogs are really cute. However, the rating denominator has to conform to some schema so we can reliably analyze these ratings later. This is why I am deeming any rating denominator value other than 10 to be invalid.*

- Erroneous data type for the `timestamp` column

**Tidiness Issues**

## Image predictions table

### Visual assessment

In [18]:
img_pred_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

<IPython.core.display.Javascript object>

In [19]:
img_pred_df.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

<IPython.core.display.Javascript object>

In [20]:
img_pred_df.sample(5)

tweet_id                                            jpg_url  \
360   672640509974827008    https://pbs.twimg.com/media/CVWy9v-VAAALSoE.jpg   
1833  836753516572119041    https://pbs.twimg.com/media/C5y-4VwWcAIcaoj.jpg   
1383  765669560888528897    https://pbs.twimg.com/media/CqA0XcYWAAAzltT.jpg   
1312  754482103782404096  https://pbs.twimg.com/ext_tw_video_thumb/75448...   
234   670421925039075328    https://pbs.twimg.com/media/CU3RLqfW4AE0pbA.jpg   

      img_num                        p1   p1_conf  p1_dog             p2  \
360         1  Chesapeake_Bay_retriever  0.420155    True       Cardigan   
1833        1               mortarboard  0.936882   False  academic_gown   
1383        1                    beagle  0.993333    True   Walker_hound   
1312        1                       tub  0.596796   False        bathtub   
234         1                 Chihuahua  0.275793    True           corn   

       p2_conf  p2_dog                  p3   p3_conf  p3_dog  
360   0.266030    True  Labrador_retriever  0.042514    True  
1833  0.020815   False          schipperke  0.011564    True  
1383  0.002902    True              basset  0.002415    True  
1312  0.381098   False      shower_curtain  0.017629   False  
234   0.073596   False              bolete  0.054905   False

<IPython.core.display.Javascript object>

### Programmatic assessment

In [21]:
img_pred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


<IPython.core.display.Javascript object>

In [22]:
img_pred_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

<IPython.core.display.Javascript object>

### Findings

This table looks much cleaner which is something I anticipated, it was created by a data analyst after all. 

However, ther are minor inconveniences which must be mentioned.

- Missing records compared to tweet_archive_df, this table has 2075 entry while the former one has 2356 entries.
- Entries of images that do not have a dog in them i.e. `p1_dog` = `p2_dog` = `p3_dog` = False